### PR TITLE
release: add installer validation

### PR DESCRIPTION
**Pull request by ldennington**
Add basic installer validation to release pipeline for Windows,
macOS (x64 only), and Linux (Debian package only). Validation runs the
installers/any necessary setup and `git-credential-manager-core
--version`.

Test run validating these changes can be found [here](https://github.com/ldennington/git-credential-manager/actions/runs/2757819748).

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -437,13 +437,53 @@ jobs:
...
 
+# ================================
+#           Validate
+# ================================
+  validate:
+    name: Validate installers
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, macos-latest, windows-latest ]
```
Comment (dscho) [inline .github/workflows/release.yml:448]: This looks great!

One suggestion: It might make sense to imitate the matrix style of [`monitor-components.yml`](https://github.com/git-for-windows/git/blob/79972f5b45e66439383630844fb843a84ccf5cce/.github/workflows/monitor-components.yml#L23) where the matrix values specify more than a single feature. That way, we could express the `needs` value in terms of a matrix variable, and we could likewise avoid downloading unneeded artifacts, again by using a matrix variable.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -437,13 +437,53 @@ jobs:
...
+    needs: [ osx-sign, win-sign, linux-sign ]
+    steps:
+      - name: Download artifacts
+        uses: actions/download-artifact@v3
```
Comment (dscho) [inline .github/workflows/release.yml:452]: We could restrict this via `uses: name: <matrix-variable>`

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -437,13 +437,53 @@ jobs:
...
+
+      - name: Build & Install Windows
+        if: contains(matrix.os, 'windows')
+        shell: sh
```
Comment (dscho) [inline .github/workflows/release.yml:456]: Oh, that's allowed? I thought the permissible value was called `bash`...

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -437,13 +437,53 @@ jobs:
...
+        run: |
+          exePaths=$(find ./win-sign/signed/*.exe) &&
+          for path in $exePaths; do
+            start "" "$path" /WAIT //SILENT //NORESTART &&
```
Comment (dscho) [inline .github/workflows/release.yml:460]: We probably also want `//VERYSILENT`.

**Review by dscho — APPROVED** [review]
Excellent work!

I had a few suggestions, but would also be fine with the current iteration.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -437,13 +437,53 @@ jobs:
...
+            git-credential-manager-core --version
+          done
+
+      - name: Build & Install Linux
+        if: contains(matrix.os, 'ubuntu')
+        run: |
+          debpath=$(find ./linux-sign/*.deb)
+          sudo apt install $debpath
+          git-credential-manager-core configure
+          git-credential-manager-core --version
```
Comment (derrickstolee) [inline .github/workflows/release.yml:470]: Since we know that GCM is not installed _a priori_ on the build machines, this demonstrates that the executable was installed on the `PATH`. Is there any way we could validate the output of `git-credential-manager-core --version` to be sure we have the right version number?

I see this as a really important validation for the Git installer validation, but might be helpful here.

**Review by derrickstolee — APPROVED** [review]

File: .github/workflows/release.yml (hunk 5 above)
Comment (mjcheetham) [reply .github/workflows/release.yml:470]: If we could print the output that would be great indeed! I think the script task here should do this already, no?

File: .github/workflows/release.yml (hunk 4 above)
Comment (mjcheetham) [reply .github/workflows/release.yml:460]: 👍

File: .github/workflows/release.yml (hunk 1 above)
Comment (mjcheetham) [reply .github/workflows/release.yml:448]: +1 for the matrix/vector

**Review by mjcheetham — APPROVED** [review]
This looks great! Some of the comments from the others would help debugging problems and speed things up.

File: .github/workflows/release.yml (hunk 5 above)
Comment (derrickstolee) [reply .github/workflows/release.yml:470]: I was thinking something like

```
git-credential-manager-core --version >actual
cat >expect <<-EOF
$GCM_VERSION_NUMBER
EOF
cmp expect actual || return 1
```

so we get an automated build failure if they are different.

File: .github/workflows/release.yml (hunk 1 above)
Comment (ldennington) [reply .github/workflows/release.yml:448]: This is great!

File: .github/workflows/release.yml (hunk 3 above)
Comment (ldennington) [reply .github/workflows/release.yml:456]: It does seem to be allowed, but I can update it to bash.

File: .github/workflows/release.yml (hunk 4 above)
Comment (ldennington) [reply .github/workflows/release.yml:460]: To clarify - do we want `//VERYSILENT` in addition to `//SILENT` or instead of `//SILENT`?

Comment (mjcheetham) [reply .github/workflows/release.yml:460]: I think `//VERYSILENT` includes `//SILENT`, but not 100% sure. Docs don't make it too clear:

https://jrsoftware.org/ishelp/topic_setupcmdline.htm#SILENT

Comment (dscho) [reply .github/workflows/release.yml:460]: In Git for Windows, we always use `//VERYSILENT` _in addition to_ `//SILENT`, but it might be a "better safe 'n sorry" thing. I have no idea whether `//VERYSILENT` _implies_ `//SILENT`. I guess it's time to consult the source.

*clicketyclick*

Seems that [`//VERYSILENT` sets the mode to something different than `//SILENT`](https://github.com/jrsoftware/issrc/blob/cb9c3e94593724ea062bb3070bb034abdd009652/Projects/Main.pas#L2781-L2784), but it would appear as if `//VERYSILENT` indeed does imply `//SILENT` (there is only [this](https://github.com/jrsoftware/issrc/blob/cb9c3e94593724ea062bb3070bb034abdd009652/Projects/Main.pas#L2786-L2792) and [this](https://github.com/jrsoftware/issrc/blob/cb9c3e94593724ea062bb3070bb034abdd009652/Projects/Main.pas#L3069-L3070) and [this](https://github.com/jrsoftware/issrc/blob/cb9c3e94593724ea062bb3070bb034abdd009652/Projects/Main.pas#L3265) and [this](https://github.com/jrsoftware/issrc/blob/cb9c3e94593724ea062bb3070bb034abdd009652/Projects/Main.pas#L4267-L4278) part of the code, and it all looks as if the `imVerySilent` mode does more than `imSilent`, not different things).

tl;dr it does seem as if `//VERYSILENT` should be enough. To be safe, though, I would use both (I vaguely remember that we had to do this at some stage, probably due to a hence-fixed bug in InnoSetup).

Comment (ldennington) [reply .github/workflows/release.yml:460]: Okie doke, I'll add both!

**Comment by ldennington** [comment]
I believe we are all good on this! Here is the latest successful test run: https://github.com/ldennington/git-credential-manager/actions/runs/2852519783.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -437,13 +437,81 @@ jobs:
...
+          # Only validate x64, given arm64 agents are not available
+          pkgpath=$(find ./*.pkg)
+          sudo installer -pkg $pkgpath -target /
+
+      - name: Validate
+        shell: bash
+        run: |
+          "${{ matrix.component.command }}" --version | sed 's/+.*//' >actual
+          echo $GitBuildVersionSimple >expect
+          cmp expect actual || exit 1
```
Comment (derrickstolee) [inline .github/workflows/release.yml:506]: Very nice!

**Review by derrickstolee — APPROVED** [review]

**Comment by Favourjacobmudiaga** [comment]
Alright
